File: Frontend-1/src/api/welfare.ts
import { api } from "./client";
import type { PaginatedResponse, MessageResponse, Member } from "./types";

// --- Types ---

export type WelfareEventType =
  | "MSIBA"
  | "HARUSI"
  | "DHARURA"
  | "MATIBABU"
  | "KUZALIWA"
  | "ELIMU";

export type WelfareFundingSource =
  | "TREASURY"
  | "MEMBER_CONTRIBUTION"
  | "BOTH";

export type WelfareEventStatus =
  | "PENDING"
  | "APPROVED"
  | "REJECTED"
  | "COMPLETED";

export type WelfareContributionStatus =
  | "PENDING"
  | "PAID"
  | "WAIVED";

export interface WelfareEvent {
  id: string;
  member_id: string;
  event_type: WelfareEventType;
  description: string;
  amount_requested: number;
  amount_approved?: number;
  funding_source: WelfareFundingSource;
  treasury_amount: number;
  member_amount: number;
  status: WelfareEventStatus;
  created_by: string;
  approved_by?: string;
  rejected_by?: string;
  rejection_reason?: string;
  completed_at?: string;
  created_at: string;
  updated_at: string;
  member?: Pick<Member, "id" | "member_no" | "full_name" | "phone">;
  creator?: { id: string; name: string; role: string };
  approver?: { id: string; name: string; role: string };
}

export interface WelfareContribution {
  id: string;
  event_id: string;
  member_id: string;
  amount: number;
  status: WelfareContributionStatus;
  paid_at?: string;
  recorded_by?: string;
  created_at: string;
  updated_at: string;
  event?: Pick<WelfareEvent, "id" | "event_type" | "description" | "status">;
  member?: Pick<Member, "id" | "member_no" | "full_name" | "phone">;
  recorder?: { id: string; name: string; role: string };
}

export interface WelfareDashboard {
  total_events: number;
  pending_approval: number;
  active_events: number;
  completed_events: number;
  rejected_events: number;
  total_collected: number;
  total_from_treasury: number;
  my_pending_contributions: number;
  my_paid_contributions: number;
}

export interface CreateWelfareEventRequest {
  member_id: string;
  event_type: WelfareEventType;
  description: string;
  amount_requested: number;
  funding_source: WelfareFundingSource;
  treasury_amount: number;
  member_amount: number;
}

export interface ApproveWelfareEventRequest {
  approved_amount: number;
}

export interface RejectWelfareEventRequest {
  reason: string;
}

export interface RecordWelfarePaymentRequest {
  amount: number;
}

export interface WelfareEventDetail {
  data: WelfareEvent;
  contributions: WelfareContribution[];
  stats: {
    total_paid: number;
    total_pending: number;
    paid_count: number;
    pending_count: number;
  };
}

// --- API ---

export const welfareApi = {
  // Dashboard
  getDashboard: () =>
    api.get<{ data: WelfareDashboard }>("/welfare/dashboard"),

  // Events
  listEvents: (params?: {
    page?: number;
    limit?: number;
    status?: string;
    event_type?: string;
  }) => {
    const q: Record<string, string> = {};
    if (params?.page) q.page = String(params.page);
    if (params?.limit) q.limit = String(params.limit);
    if (params?.status) q.status = params.status;
    if (params?.event_type) q.event_type = params.event_type;
    return api.get<PaginatedResponse<WelfareEvent>>("/welfare/events", q);
  },

  getEvent: (id: string) =>
    api.get<WelfareEventDetail>(`/welfare/events/${id}`),

  createEvent: (data: CreateWelfareEventRequest) =>
    api.post<{ message: string; data: WelfareEvent }>("/welfare/events", data),

  approveEvent: (id: string, data: ApproveWelfareEventRequest) =>
    api.post<{ message: string; data: WelfareEvent }>(
      `/welfare/events/${id}/approve`,
      data
    ),

  rejectEvent: (id: string, data: RejectWelfareEventRequest) =>
    api.post<{ message: string; data: WelfareEvent }>(
      `/welfare/events/${id}/reject`,
      data
    ),

  // Contributions
  listContributions: (params?: {
    page?: number;
    limit?: number;
    status?: string;
    event_id?: string;
  }) => {
    const q: Record<string, string> = {};
    if (params?.page) q.page = String(params.page);
    if (params?.limit) q.limit = String(params.limit);
    if (params?.status) q.status = params.status;
    if (params?.event_id) q.event_id = params.event_id;
    return api.get<PaginatedResponse<WelfareContribution>>(
      "/welfare/contributions",
      q
    );
  },

  myContributions: (params?: {
    page?: number;
    limit?: number;
    status?: string;
  }) => {
    const q: Record<string, string> = {};
    if (params?.page) q.page = String(params.page);
    if (params?.limit) q.limit = String(params.limit);
    if (params?.status) q.status = params.status;
    return api.get<PaginatedResponse<WelfareContribution>>(
      "/welfare/my-contributions",
      q
    );
  },

  recordPayment: (
    eventId: string,
    memberId: string,
    data: RecordWelfarePaymentRequest
  ) =>
    api.post<MessageResponse>(
      `/welfare/events/${eventId}/contributions/${memberId}/pay`,
      data
    ),

  waiveContribution: (eventId: string, memberId: string) =>
    api.post<MessageResponse>(
      `/welfare/events/${eventId}/contributions/${memberId}/waive`
    ),
};

File: backend/models/welfare.go
package models

import (
	"time"

	"github.com/shopspring/decimal"
)

// WelfareEventType represents the type of welfare event.
type WelfareEventType string

const (
	WelfareMisiba   WelfareEventType = "MSIBA"    // Death/Funeral
	WelfareHarusi   WelfareEventType = "HARUSI"   // Wedding
	WelfareDharura  WelfareEventType = "DHARURA"  // Emergency
	WelfareMedical  WelfareEventType = "MATIBABU" // Medical
	WelfareKuzaliwa WelfareEventType = "KUZALIWA" // Birth
	WelfareElimu    WelfareEventType = "ELIMU"    // Education
)

// WelfareFundingSource represents how the welfare event is funded.
type WelfareFundingSource string

const (
	FundTreasury            WelfareFundingSource = "TREASURY"
	FundMemberContribution  WelfareFundingSource = "MEMBER_CONTRIBUTION"
	FundBoth                WelfareFundingSource = "BOTH"
)

// WelfareEventStatus represents the lifecycle status of a welfare event.
type WelfareEventStatus string

const (
	WelfarePending   WelfareEventStatus = "PENDING"
	WelfareApproved  WelfareEventStatus = "APPROVED"
	WelfareRejected  WelfareEventStatus = "REJECTED"
	WelfareCompleted WelfareEventStatus = "COMPLETED"
)

// WelfareEvent represents a welfare event (misiba, harusi, dharura, etc.).
type WelfareEvent struct {
	ID              string               `gorm:"primaryKey;type:uuid;default:gen_random_uuid()" json:"id"`
	MemberID        string               `gorm:"type:uuid;not null;index" json:"member_id"`
	EventType       WelfareEventType     `gorm:"type:varchar(20);not null" json:"event_type"`
	Description     string               `gorm:"type:text;not null" json:"description"`
	AmountRequested decimal.Decimal      `gorm:"type:decimal(15,2);not null" json:"amount_requested"`
	AmountApproved  *decimal.Decimal     `gorm:"type:decimal(15,2)" json:"amount_approved,omitempty"`
	FundingSource   WelfareFundingSource `gorm:"type:varchar(30);not null" json:"funding_source"`
	TreasuryAmount  decimal.Decimal      `gorm:"type:decimal(15,2);not null;default:0" json:"treasury_amount"`
	MemberAmount    decimal.Decimal      `gorm:"type:decimal(15,2);not null;default:0" json:"member_amount"`
	Status          WelfareEventStatus   `gorm:"type:varchar(20);not null;default:'PENDING';index" json:"status"`
	CreatedBy       string               `gorm:"type:uuid;not null" json:"created_by"`
	ApprovedBy      *string              `gorm:"type:uuid" json:"approved_by,omitempty"`
	RejectedBy      *string              `gorm:"type:uuid" json:"rejected_by,omitempty"`
	RejectionReason *string              `gorm:"type:text" json:"rejection_reason,omitempty"`
	CompletedAt     *time.Time           `json:"completed_at,omitempty"`
	DisbursedBy     *string              `gorm:"type:uuid" json:"disbursed_by,omitempty"`
	DisbursedAt     *time.Time           `json:"disbursed_at,omitempty"`
	ReceivedAt      *time.Time           `json:"received_at,omitempty"`
	ReceivedBy      *string              `gorm:"type:uuid" json:"received_by,omitempty"`
	CreatedAt       time.Time            `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt       time.Time            `gorm:"autoUpdateTime" json:"updated_at"`

	Member   *Member `gorm:"foreignKey:MemberID" json:"member,omitempty"`
	Creator  *User   `gorm:"foreignKey:CreatedBy" json:"creator,omitempty"`
	Approver *User   `gorm:"foreignKey:ApprovedBy" json:"approver,omitempty"`
	Rejector *User   `gorm:"foreignKey:RejectedBy" json:"rejector,omitempty"`
}

// WelfareContributionStatus represents the payment status of a member's welfare contribution.
type WelfareContributionStatus string

const (
	WelfareContribPending WelfareContributionStatus = "PENDING"
	WelfareContribPaid    WelfareContributionStatus = "PAID"
	WelfareContribWaived  WelfareContributionStatus = "WAIVED"
)

// WelfareContribution represents a member's obligation to contribute to a welfare event.
type WelfareContribution struct {
	ID         string                    `gorm:"primaryKey;type:uuid;default:gen_random_uuid()" json:"id"`
	EventID    string                    `gorm:"type:uuid;not null;index:idx_event_member,unique" json:"event_id"`
	MemberID   string                    `gorm:"type:uuid;not null;index:idx_event_member,unique" json:"member_id"`
	Amount     decimal.Decimal           `gorm:"type:decimal(15,2);not null" json:"amount"`
	Status     WelfareContributionStatus `gorm:"type:varchar(20);not null;default:'PENDING'" json:"status"`
	PaidAt     *time.Time                `json:"paid_at,omitempty"`
	RecordedBy *string                   `gorm:"type:uuid" json:"recorded_by,omitempty"`
	CreatedAt  time.Time                 `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt  time.Time                 `gorm:"autoUpdateTime" json:"updated_at"`

	Event    *WelfareEvent `gorm:"foreignKey:EventID" json:"event,omitempty"`
	Member   *Member       `gorm:"foreignKey:MemberID" json:"member,omitempty"`
	Recorder *User         `gorm:"foreignKey:RecordedBy" json:"recorder,omitempty"`
}
